File: go-flood/pkg/tilings/elongated_triangular.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateElongatedTriangularBoard(options Options) core.Board {
	a := options.TileSize
	cols, rows := options.Cols, options.Rows
	colorCount := options.ColorCount
	rng := options.RNG

	h := a * math.Sqrt(3.0) / 2.0
	numSquareRows := rows
	numTriangleRows := rows - 1
	totalStrips := numSquareRows + numTriangleRows

	idCounter := 0

	type vertexInfo struct {
		id int
		p  core.Point
	}
	vertexMap := make(map[string]vertexInfo)
	vertexEdgeKey := func(a, b int) string {
		if a < b {
			return fmt.Sprintf("%d,%d", a, b)
		}
		return fmt.Sprintf("%d,%d", b, a)
	}
	getVertexId := func(p core.Point) int {
		key := fmt.Sprintf("%.3f,%.3f", p[0], p[1])
		if info, ok := vertexMap[key]; ok {
			return info.id
		}
		id := len(vertexMap)
		vertexMap[key] = vertexInfo{id, p}
		return id
	}

	type vTile struct {
		points []core.Point
		vIds   []int
		tType  string
	}
	var rawVTiles []vTile

	for s := 0; s < totalStrips; s++ {
		isSquareStrip := (s%2 == 0)
		stripIdx := s / 2
		yBase := float64(stripIdx)*(a+h) + func() float64 {
			if isSquareStrip {
				return 0
			}
			return a
		}()

		currentSquareOffset := func() float64 {
			if stripIdx%2 == 0 {
				return 0
			}
			return a / 2.0
		}()
		nextSquareOffset := func() float64 {
			if (stripIdx+1)%2 == 0 {
				return 0
			}
			return a / 2.0
		}()

		if isSquareStrip {
			for q := 0; q < cols; q++ {
				x := float64(q)*a + currentSquareOffset
				y := yBase
				points := []core.Point{
					{x, y}, {x + a, y}, {x + a, y + a}, {x, y + a},
				}
				vIds := make([]int, 4)
				for i, p := range points {
					vIds[i] = getVertexId(p)
				}
				rawVTiles = append(rawVTiles, vTile{points, vIds, "square"})
			}
		} else {
			offsetBelow := currentSquareOffset
			offsetAbove := nextSquareOffset
			for q := -1; q < cols+1; q++ {
				xBelow := float64(q)*a + offsetBelow
				xAbove := float64(q)*a + offsetAbove

				apexUpX := func() float64 {
					if offsetBelow < offsetAbove {
						return xAbove
					}
					return xAbove + a
				}()
				pUp := []core.Point{
					{xBelow, yBase}, {xBelow + a, yBase}, {apexUpX, yBase + h},
				}

				apexDownX := func() float64 {
					if offsetBelow < offsetAbove {
						return xBelow + a
					}
					return xBelow
				}()
				pDown := []core.Point{
					{xAbove, yBase + h}, {xAbove + a, yBase + h}, {apexDownX, yBase},
				}

				for _, points := range [][]core.Point{pUp, pDown} {
					vIds := make([]int, 3)
					for i, p := range points {
						vIds[i] = getVertexId(p)
					}
					rawVTiles = append(rawVTiles, vTile{points, vIds, "triangle"})
				}
			}
		}
	}

	squareEdgeKeys := make(map[string]bool)
	for _, rt := range rawVTiles {
		if rt.tType != "square" {
			continue
		}
		for i := range rt.vIds {
			squareEdgeKeys[vertexEdgeKey(rt.vIds[i], rt.vIds[(i+1)%len(rt.vIds)])] = true
		}
	}
	sharesSquareEdge := func(vIds []int) bool {
		return squareEdgeKeys[vertexEdgeKey(vIds[0], vIds[1])] ||
			squareEdgeKeys[vertexEdgeKey(vIds[1], vIds[2])] ||
			squareEdgeKeys[vertexEdgeKey(vIds[2], vIds[0])]
	}

	var filteredTiles []core.Tile
	type filteredVTile struct {
		vIds []int
	}
	var filteredVTiles []filteredVTile

	for _, rt := range rawVTiles {
		if rt.tType == "square" || sharesSquareEdge(rt.vIds) {
			id := idCounter
			idCounter++
			filteredTiles = append(filteredTiles, core.Tile{
				ID:        id,
				ColorID:   int(rng() * float64(colorCount)),
				OwnerID:   nil,
				Points:    rt.points,
				Neighbors: []int{},
			})
			filteredVTiles = append(filteredVTiles, filteredVTile{rt.vIds})
		}
	}

	vertexToTiles := make([][]int, len(vertexMap))
	for tileIdx, fvt := range filteredVTiles {
		for _, vId := range fvt.vIds {
			vertexToTiles[vId] = append(vertexToTiles[vId], tileIdx)
		}
	}

	for i := range filteredTiles {
		neighborCounts := make(map[int]int)
		for _, vId := range filteredVTiles[i].vIds {
			for _, j := range vertexToTiles[vId] {
				if i == j {
					continue
				}
				neighborCounts[j]++
			}
		}
		for j, count := range neighborCounts {
			if count >= 2 {
				filteredTiles[i].Neighbors = append(filteredTiles[i].Neighbors, filteredTiles[j].ID)
			}
		}
	}

	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64
	for _, t := range filteredTiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	for i := range filteredTiles {
		for j := range filteredTiles[i].Points {
			filteredTiles[i].Points[j][0] -= minX
			filteredTiles[i].Points[j][1] -= minY
		}
	}

	findClosest := func(tx, ty float64) int {
		bestId := 0
		minDist := math.MaxFloat64
		for _, t := range filteredTiles {
			var cx, cy float64
			for _, p := range t.Points {
				cx += p[0]
				cy += p[1]
			}
			cx /= float64(len(t.Points))
			cy /= float64(len(t.Points))
			d := math.Pow(cx-tx, 2) + math.Pow(cy-ty, 2)
			if d < minDist {
				minDist = d
				bestId = t.ID
			}
		}
		return bestId
	}

	w, hBoard := maxX-minX, maxY-minY
	startTileIds := []int{
		findClosest(0, 0),
		findClosest(w, 0),
		findClosest(0, hBoard),
		findClosest(w, hBoard),
	}

	return core.Board{
		Version:      1,
		Generator:    "elongated-triangular",
		Width:        w,
		Height:       hBoard,
		Cols:         cols,
		Rows:         rows,
		Tiles:        filteredTiles,
		StartTileIds: startTileIds,
	}
}

File: js/tilings/elongated-triangular.js
/**
 * Generates an Elongated Triangular tiling board (3.3.3.4.4).
 * This tiling consists of alternating strips of squares and triangles.
 * Every vertex is shared by two squares and three triangles.
 */
export function generateElongatedTriangularBoard(options) {
  const { cols, rows, tileSize: a, colorCount, rng } = options;

  const h = a * Math.sqrt(3) / 2;

  // We want 'rows' rows of squares, with triangle strips in between.
  // Total number of strips = rows (squares) + (rows - 1) (triangles)
  const numSquareRows = rows;
  const numTriangleRows = rows - 1;
  const totalStrips = numSquareRows + numTriangleRows;

  let tiles = [];
  let idCounter = 0;
  const vertexMap = new Map();
  const vertexEdgeKey = (a, b) => a < b ? `${a},${b}` : `${b},${a}`;
  const getVertexId = (p) => {
    const key = `${p[0].toFixed(3)},${p[1].toFixed(3)}`;
    if (vertexMap.has(key)) return vertexMap.get(key).id;
    const id = vertexMap.size;
    vertexMap.set(key, { id, p });
    return id;
  };

  const squareMap = new Map();

  for (let s = 0; s < totalStrips; s++) {
    const isSquareStrip = (s % 2 === 0);
    const stripIdx = Math.floor(s / 2);

    // yBase calculation:
    // s=0 (S): 0
    // s=1 (T): a
    // s=2 (S): a + h
    // s=3 (T): 2a + h
    // ...
    const yBase = stripIdx * (a + h) + (isSquareStrip ? 0 : a);
    const offset = (stripIdx % 2 === 1 && isSquareStrip) ? a / 2 :
                   (stripIdx % 2 === 0 && !isSquareStrip) ? 0 :
                   (stripIdx % 2 === 1 && !isSquareStrip) ? a / 2 : 0;

    // Actually simpler:
    // Square Row 0 (strip 0): offset 0
    // Square Row 1 (strip 2): offset a/2
    // Square Row 2 (strip 4): offset 0
    const currentSquareOffset = (stripIdx % 2 === 0) ? 0 : a / 2;
    const nextSquareOffset = ((stripIdx + 1) % 2 === 0) ? 0 : a / 2;

    if (isSquareStrip) {
      for (let q = 0; q < cols; q++) {
        const x = q * a + currentSquareOffset;
        const y = yBase;
        const points = [
          [x, y],
          [x + a, y],
          [x + a, y + a],
          [x, y + a]
        ];
        const vIds = points.map(getVertexId);
        const tile = {
          id: idCounter++,
          colorId: Math.floor(rng() * colorCount),
          ownerId: null,
          points,
          vIds,
          type: 'square',
          neighbors: []
        };
        tiles.push(tile);
        squareMap.set(`${q},${stripIdx}`, tile.id);
      }
    } else {
      // Triangles between Square strip 'stripIdx' and 'stripIdx + 1'
      const offsetBelow = currentSquareOffset;
      const offsetAbove = nextSquareOffset;

      for (let q = -1; q < cols + 1; q++) {
        const xBelow = q * a + offsetBelow;
        const xAbove = q * a + offsetAbove;

        // Up triangle
        // If offsetBelow=0, offsetAbove=0.5a: vertices at yBase: q*a, (q+1)*a. Apex at yBase+h: (q+0.5)a
        // If offsetBelow=0.5a, offsetAbove=0: vertices at yBase: (q+0.5)a, (q+1.5)a. Apex at yBase+h: (q+1)a
        const apexUpX = (offsetBelow < offsetAbove) ? xAbove : xAbove + a;
        const pUp = [
          [xBelow, yBase],
          [xBelow + a, yBase],
          [apexUpX, yBase + h]
        ];

        // Down triangle
        // If offsetBelow=0, offsetAbove=0.5a: vertices at yBase+h: (q+0.5)a, (q+1.5)a. Apex at yBase: (q+1)a
        // If offsetBelow=0.5a, offsetAbove=0: vertices at yBase+h: q*a, (q+1)a. Apex at yBase: (q+0.5)a
        const apexDownX = (offsetBelow < offsetAbove) ? xBelow + a : xBelow;
        const pDown = [
          [xAbove, yBase + h],
          [xAbove + a, yBase + h],
          [apexDownX, yBase]
        ];

        [pUp, pDown].forEach(points => {
          tiles.push({
            id: idCounter++,
            colorId: Math.floor(rng() * colorCount),
            ownerId: null,
            points,
            vIds: points.map(getVertexId),
            type: 'triangle',
            neighbors: []
          });
        });
      }
    }
  }

  const squareEdgeKeys = new Set();
  tiles.filter(t => t.type === 'square').forEach(tile => {
    for (let i = 0; i < tile.vIds.length; i++) {
      squareEdgeKeys.add(vertexEdgeKey(tile.vIds[i], tile.vIds[(i + 1) % tile.vIds.length]));
    }
  });

  const sharesSquareEdge = (tile) => (
    squareEdgeKeys.has(vertexEdgeKey(tile.vIds[0], tile.vIds[1])) ||
    squareEdgeKeys.has(vertexEdgeKey(tile.vIds[1], tile.vIds[2])) ||
    squareEdgeKeys.has(vertexEdgeKey(tile.vIds[2], tile.vIds[0]))
  );

  const filteredTiles = tiles.filter(t => t.type === 'square' || sharesSquareEdge(t));

  // Re-index
  filteredTiles.forEach((t, idx) => t.id = idx);

  // Connectivity via vIds
  const vertexToTiles = new Map();
  filteredTiles.forEach((tile, idx) => {
    tile.vIds.forEach(vId => {
      if (!vertexToTiles.has(vId)) vertexToTiles.set(vId, []);
      vertexToTiles.get(vId).push(idx);
    });
  });

  filteredTiles.forEach((tile, i) => {
    const neighborCounts = new Map();
    tile.vIds.forEach(vId => {
      vertexToTiles.get(vId).forEach(j => {
        if (i === j) return;
        neighborCounts.set(j, (neighborCounts.get(j) || 0) + 1);
      });
    });
    neighborCounts.forEach((count, j) => {
      if (count >= 2) {
        tile.neighbors.push(filteredTiles[j].id);
      }
    });
  });

  // Final bounding box and normalization
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;
  for (const t of filteredTiles) {
    for (const p of t.points) {
      minX = Math.min(minX, p[0]);
      minY = Math.min(minY, p[1]);
      maxX = Math.max(maxX, p[0]);
      maxY = Math.max(maxY, p[1]);
    }
  }

  filteredTiles.forEach(t => {
    t.points = t.points.map(p => [p[0] - minX, p[1] - minY]);
  });

  const findClosest = (tx, ty) => {
    let bestId = 0;
    let minDist = Infinity;
    filteredTiles.forEach(t => {
      const cx = t.points.reduce((sum, p) => sum + p[0], 0) / t.points.length;
      const cy = t.points.reduce((sum, p) => sum + p[1], 0) / t.points.length;
      const d = Math.pow(cx - tx, 2) + Math.pow(cy - ty, 2);
      if (d < minDist) {
        minDist = d;
        bestId = t.id;
      }
    });
    return bestId;
  };

  const width = maxX - minX;
  const height = maxY - minY;
  const startTileIds = [
    findClosest(0, 0),
    findClosest(width, 0),
    findClosest(0, height),
    findClosest(width, height)
  ];

  return {
    version: 1,
    generator: "elongated-triangular",
    width,
    height,
    tiles: filteredTiles,
    startTileIds
  };
}
